// @mocypress/mocypress-ui-core/src/main/behaviours/Clickable.ts
import {BaseBehaviour} from "../components/generalize/BaseBehaviour";
import {Method} from "../components/generalize/Method";

/**
 * A library for building mocypress-ui-core dependency lib
 *
 * @remarks
 * The mocypress-ui-core defines the {@link Clickable} class and {@link BaseBehaviour} abstract class,
 * which are used to build element clickable behaviours
 *
 * @packageDocumentation
 */

/**
 * Class implemented by to re-present all generalize clickable behaviours for web-components
 * @public
 */
export class Clickable extends BaseBehaviour{

    constructor(locator: string, method: Method) {
        super(locator,method);
    }

    public click(): void{
        this.element.click();
    }

    public click_force(): void{
        this.element.click({force: true})
    }

    public click_position(position: any): void{
        this.element.click(position)
    }

    public click_double(): void {
        this.element.dblclick()
    }

    public click_right(): void {
        this.element.rightclick()
    }

}
